File: src/leetcode/constructbtfrompreorderinorder/ConstructBinaryTreeFromPreorderAndInorderTraversal.java
package leetcode.constructbtfrompreorderinorder;

import trees.TreeNode;

import java.util.HashMap;

public class ConstructBinaryTreeFromPreorderAndInorderTraversal {

    int preOrderPointer = 0;

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        HashMap<Integer, Integer> inOrderNodeToIndex = new HashMap<>();
        for (int i = 0; i < inorder.length; i++) {
            inOrderNodeToIndex.put(inorder[i], i);
        }

        // preorder[0] is the root of the entire tree
        return construct(preorder, 0, inorder.length - 1, inOrderNodeToIndex);

    }

    private TreeNode construct(int[] preorder,
                               int nodeRangeStartInInOrderRA,
                               int nodeRangeEndInInOrderRA,
                               HashMap<Integer, Integer> inOrderNodeToIndex) {
        if (nodeRangeStartInInOrderRA <= nodeRangeEndInInOrderRA) {
            int rootVal = preorder[preOrderPointer];
            Integer rootIndexInInOrderRA = inOrderNodeToIndex.get(rootVal);
            TreeNode root = new TreeNode(rootVal);
            preOrderPointer++;
            if (nodeRangeStartInInOrderRA < nodeRangeEndInInOrderRA) {
                root.left = construct(preorder, nodeRangeStartInInOrderRA, rootIndexInInOrderRA - 1, inOrderNodeToIndex);
                root.right = construct(preorder, rootIndexInInOrderRA + 1, nodeRangeEndInInOrderRA, inOrderNodeToIndex);
            }
            return root;
        } else {
            return null;
        }
    }
}

File: python-code/construct_binary_tree_from_pre_order_and_in_order_traversal.py
from typing import *
from tree_node import TreeNode


class Solution:
    """
    preorder = root,left,right
    inorder = left,root,right
    """

    def construct_tree_from_preorder_and_inorder(
        self, preorder: List[int], inorder: List[int]
    ) -> Optional[TreeNode]:
        if not preorder or not inorder:
            return None

        root_val: int = preorder.pop(0)
        root: TreeNode = TreeNode(root_val)
        inorder_root_val_index = inorder.index(root_val)
        root.left = self.construct_tree_from_preorder_and_inorder(
            preorder, inorder[:inorder_root_val_index]
        )
        root.right = self.construct_tree_from_preorder_and_inorder(
            preorder, inorder[inorder_root_val_index + 1 :]
        )

        return root


if __name__ == "__main__":
    reconstructed_tree = Solution().construct_tree_from_preorder_and_inorder(
        [3, 9, 20, 15, 7], [9, 3, 15, 20, 7]
    )
    print(reconstructed_tree)
